pipeline_branching.md

```diff
@@ -0,0 +1,159 @@
+# Pipeline Branching
+
+## Goal
+
+Allow users to have multiple, named versions of a pipeline (branches)
+running concurrently.
+
+## Use Case:
+
+The primary use case for this is to give users a good way to have `dev`,
+`staging` and `prod` versions of pipelines. There are a ton of use cases for
+this feature, but so far I haven't encountered one that presented challenges
+that the `dev`, `staging` `prod` use case doesn't, so for this document we'll
+just stick to that for inspiration.
+
+## PPS API changes:
+The pipeline message will be extended to include a `Branch` field, which
+is a string:
+
+```proto
+message Pipeline {
+  string name = 1;
+  string branch = 2;
+}
+```
+
+In pachctl this is implemented with the familiar syntax of `pipeline@branch`.
```

Comment:
Are we overloading the use of @branch to support this new use case?

Reply:
It's sort of overloading, we're reusing a syntax that's used elsewhere, in terms of implementation it's not overloading though because that syntax only applied to repos, so there won't be any ambiguities we have to deal with in the code. I think the overloading of the syntax will hopefully make it easier for users to grok, since it has the same meaning in both cases.